### DISCUSSION TITLE
Planning Upcoming Tests with APS Algorithm: Resource Allocation Challenge

### DISCUSSION THREAD

**Discussion by Student156** [frePPLe/frepple#672] [Q&A]
I'm developing a system to plan upcoming tests using the APS (Advanced Planning and Scheduling) algorithm. This is not for production use, but rather a test environment for planning future tests.

### Operation Structure
I've set up an operation as shown in the following image:
![image](https://github.com/user-attachments/assets/0e8c43fb-6365-4782-80fc-225659f04c55)

The process involves:

1. Combining 2-3 incoming items into a new item
2. Merging these 2 items into 1 item
3. Transferring this item into the final item

All operations between these items have zero duration, serving as placeholders for the entire process. These cannot be omitted and must be executed on the same day, just before the last operation.

### Resource Allocation Challenge
My current challenge is configuring the solver to recognize that the 4 resources in the image:

1. Have fixed working times
2. Can only perform one operation at a time

### Proposed Resource Table
To address this, I've structured the Resource Table as follows:
![image](https://github.com/user-attachments/assets/241c332e-b6c8-44a9-bed3-14ac2dd66ba4)

The "Arbeitszeiten" (Working Hours) Calendar is defined as:
![image](https://github.com/user-attachments/assets/80619fdf-7514-47e9-8b28-c1ef382e921f)
![image](https://github.com/user-attachments/assets/4fd5d413-9a6d-45cf-9d42-b4fee6491154)

### Current Results and Issues
When I run the solver for a sales order (the last item in the first image), it appears to work. However, in the "Plan Editor", it displays as follows:
![image](https://github.com/user-attachments/assets/a9912cc6-4e03-4754-9ef6-4e1107b0e216)

Clicking on the operation reveals that no resource is planned, and the resources have no assigned operations. Despite this, all Manufacturing Orders are created:
![image](https://github.com/user-attachments/assets/0a61cd0f-c3de-4f20-91da-f497cd28163f)

I've extensively searched for a solution or similar examples in the documentation and GitHub, but haven't found a resolution. Any insights or suggestions would be greatly appreciated.

**Comment by Student156** [comment]
Perhaps I simply don't understand how to manage resource availability for the solver, even after reviewing the documentation. I believe I need a procedure that combines 'We have N hours available per week' with the 'Arbeitszeiten' (Working Hours) calendar to define when the resource is actually available, while also ensuring that the resource can only handle one task at a time.

**Comment by jdetaeye** [comment]
>> also ensuring that the resource can only handle one task at a time.

Use resources of type "default".
Use an "available" or a "maximum" calendar to define their working hours.

But if the operation duration is 0, then doing one task at a time is meaningless.  The resource can do 1 task, and will be available again for a new task.... after 0 seconds. 🤔 

The supply indicates that you have correctly linked the operations and the resources.   Proposed manufacturing orders should then be assigned to the child resources.
If a data issue (such as skills or effectivity dates) causes trouble you should be able to see a message in the log file.

**Comment by Student156** [comment]
The Point with a duration of 0 was set because I initially attempted to assign a duration to the operation; however, the solver scheduled it for the day before, which was not suitable for my case. Therefore, I adjusted it to 0 and combined the time with the last operation. Is it possible to instruct the solver to plan these two operations together without any intervening time?


Ok now i changed the resources type to "default" and i choose the "Arbeitszeiten" Calendar for the Availabe times.
![image](https://github.com/user-attachments/assets/f450fc1e-ebfb-4f5b-b8b3-f6f3fbc73e17)

When i open the plan now it shows me operations for the ressources but it plans the operations out of the "Arbeitszeiten" Calendar. 
![7](https://github.com/user-attachments/assets/3aa35b0b-bed8-4c1b-affd-521882fc74f6)
I defined working times only from 8am to 5pm and the solver plans the operation from 0am to 8am.
Thats my Problem. How can i tell the solver to plan only inside of the Workingstimes ("Arbeitszeiten")?
Thanks for the patience to help me

**Comment by jdetaeye** [comment]
Your calendar is actually wrong.
The default value is 1, and during the working hours the value is also 1.  So, 24/7 availability.
Change the default value to 0.

If that's corrected, you should see gray zones in plan editor with unavailable time.
![image](https://github.com/user-attachments/assets/cf29e9f8-0485-4e2b-b8c0-aeb61a0fc137)

**Comment by Student156** [comment]
The image now displays the grey bars as intended:
![image](https://github.com/user-attachments/assets/bf643061-2c4f-4c2f-9ed0-b4c56f7e50f4)
However, a significant issue has resurfaced: the solver is scheduling the operations across two separate days. In reality, these operations must be completed within a single day. This constraint is precisely why I aimed to have the operations proceed continuously within the designated work time.

**Comment by jdetaeye** [comment]
When using an "available" calendar, the solver is stretching the MO across unavailable period.   Ie "Start today, go home and finish tomorrow".

When using a "maximum" calendar, there is no such stretching.  The complete MO must then fit within the white zones of the plan editor. 
A common gotcha with such "max calendars": if you have an MO that needs 48 hours of work, you'll never find 48 hours uninterrupted by a gray zone. Such an MO can then not be planned.

**Comment by Student156** [comment]
I’m very pleased that the plan now reflects my intentions!
I have one more question: Is it possible to assign durations to the other operations while ensuring they are scheduled together with the last one, rather than on the previous day or earlier? If this isn’t feasible, I’m fine with leaving it as it is.

**Comment by jdetaeye** [comment]
Not possible.   The MOs are scheduled independently by the algorithm, each with it's own constraints.

**Comment by Student156** [comment]
No problem at all! I have now found a way to visualize it by allocating the time to the last operation.

Thank you so much for your help, it helped me a lot!
